osu.Server.BeatmapSubmission/DatabaseOperationExtensions.cs

```diff
@@ -397,5 +397,24 @@ public static async Task<IEnumerable<uint>> GetBeatmapOwnersAsync(this MySqlConn
                 },
                 transaction);
         }
+
+        public static async Task<IEnumerable<osu_mirror>> GetMirrorsRequiringUpdateAsync(this MySqlConnection db, MySqlTransaction? transaction = null)
+        {
+            return await db.QueryAsync<osu_mirror>(
+                "SELECT * FROM `osu_mirrors` WHERE `version` > 1 AND `perform_updates` > 0",
+                transaction: transaction);
+        }
+
+        public static async Task MarkPendingPurgeAsync(this MySqlConnection db, osu_mirror mirror, uint beatmapSetId, MySqlTransaction? transaction = null)
+        {
+            await db.ExecuteAsync(
+                "UPDATE `osu_mirrors` SET `pending_purge` = CONCAT(IFNULL(`pending_purge`, ''), @beatmapset_id) WHERE `mirror_id` = @mirror_id",
```

Comment:
I think this needs a `,` as part of the concat.

Reference:

```php
$conn->exec("UPDATE osu_mirrors SET pending_purge = CONCAT(pending_purge, '{$beatmapSetId},') WHERE mirror_id = {$m['mirror_id']}");
```

also according to production, `pending_purge` cannot be null so you should be able to drop the `IFNULL`:

```sql
-- auto-generated definition
create table osu_mirrors
(
    mirror_id        tinyint auto_increment
        primary key,
    base_url         varchar(255)                              not null,
    traffic_used     bigint(11)    default 0                   not null,
    traffic_limit    bigint(11)    default 0                   not null,
    secret_key       varchar(50)   default 'osudownloadm1rr0r' not null,
    provider_user_id int unsigned  default 0                   not null,
    enabled          tinyint       default 1                   not null,
    version          decimal(4, 2)                             null,
    pending_purge    varchar(8192) default ''                  not null,
    perform_updates  tinyint       default 1                   not null,
    regions          varchar(8192)                             null,
    disk_space_free  bigint                                    null,
    is_master        tinyint(1)    default 0                   not null
)
    charset = utf8mb3
    row_format = DYNAMIC;
```

Reply:
> I think this needs a `,` as part of the concat.

Good spot :grimacing: https://github.com/ppy/osu-server-beatmap-submission/pull/14/commits/45137e95bec40551a6c7aa84ec4a004949d33d5c should do it...?

> also according to production, `pending_purge` cannot be null so you should be able to drop the `IFNULL`:

Interesting. On all my local environments it's nullable.

```sql
CREATE TABLE `osu_mirrors` (
  `mirror_id` tinyint NOT NULL AUTO_INCREMENT,
  `base_url` varchar(255) NOT NULL,
  `traffic_used` bigint NOT NULL DEFAULT '0',
  `traffic_limit` bigint NOT NULL DEFAULT '0',
  `secret_key` varchar(50) NOT NULL DEFAULT '',
  `provider_user_id` int unsigned NOT NULL DEFAULT '0',
  `enabled` tinyint NOT NULL DEFAULT '1',
  `version` decimal(4,2) DEFAULT NULL,
  `pending_purge` varchar(6000) DEFAULT NULL,
  `perform_updates` tinyint NOT NULL DEFAULT '1',
  `regions` varchar(6000) DEFAULT NULL,
  `disk_space_free` bigint DEFAULT NULL,
  `is_master` tinyint(1) NOT NULL DEFAULT '0',
  PRIMARY KEY (`mirror_id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
```

Probably something or other with osu-web's autogen'd schema definitions not being 100% there on nullability...